src/relay/backend/build_module.cc

```diff
@@ -0,0 +1,669 @@
+};
+
+/*! \brief Relay Building configuration */
+struct RelayBuildConfig {
+  int opt_level{2};
+  std::string fall_back_device{"llvm"};
+  std::unordered_set<std::string> add_pass;
+  std::unordered_set<std::string> disabled_pass;
+  OptPassLevel  OPT_PASS_LEVEL;
+  inline bool pass_enabled(std::string pass_name) const {
+    if (disabled_pass.count(pass_name)) {
+      return false;
+    }
+    if (add_pass.count(pass_name)) {
+      return true;
+    }
+    return opt_level >= OPT_PASS_LEVEL[pass_name];
+  }
+};
+
+/*! \brief GraphCodegen wrapper */
+struct GraphCodegen {
+ public:
+  GraphCodegen() {
+    auto pf = GetPackedFunc("relay.build_module._GraphRuntimeCodegen");
+    mod = (*pf)();
+  }
+  ~GraphCodegen() {}
+
+  void Init(runtime::Module* m,
+            Map<HalideIR::Expr, HalideIR::Expr> targets) {
+    Array<HalideIR::Expr> tgts;
+    for (auto kv : targets) {
+      tgts.push_back(kv.first);
+      tgts.push_back(kv.second);
+    }
+    _CallFunc("init", m, tgts);
+  }
+
+  void Codegen(Function func) {
+    _CallFunc("codegen", func);
+  }
+
+  std::string GetJSON() {
+    return _CallFunc<std::string>("get_graph_json", nullptr);
+  }
+
+  Map<std::string, Array<LoweredFunc> > GetLoweredFunc() {
+    return _CallFunc<Map<std::string, Array<LoweredFunc> > >("get_lowered_funcs", nullptr);
+  }
+
+  std::unordered_map<std::string, tvm::runtime::NDArray> GetParams() {
+    std::unordered_map<std::string, tvm::runtime::NDArray> ret;
+    auto names = _CallFunc<Array<HalideIR::Expr> >("list_params_name", nullptr);
+    for (auto expr : names) {
+      auto key = expr.as<ir::StringImm>()->value;
+      ret[key] = _CallFunc<runtime::NDArray>("get_param_by_name", key);
+    }
+    return ret;
+  }
+
+ protected:
+  tvm::runtime::Module mod;
+  template<typename R, typename ...Args>
+  R _CallFunc(const std::string &name, Args... args) {
+    auto pf = mod.GetFunction(name, false);
+    return pf(std::forward<Args>(args)...);
+  }
+  template<typename ...Args>
+  void _CallFunc(const std::string &name, Args... args) {
+    auto pf = mod.GetFunction(name, false);
+    pf(std::forward<Args>(args)...);
+    return;
+  }
+};
+
+template<typename R, typename ...Args>
+R _CallPacked(const std::string &name, Args... args) {
+  auto pf = GetPackedFunc(name);
+  return (*pf)(std::forward<Args>(args)...);
+}
+
+template<typename ...Args>
+Function _CallPacked(const std::string &name, Args... args) {
+  auto pf = GetPackedFunc(name);
+  return (*pf)(std::forward<Args>(args)...);
+}
+
+
+class RelayBuildModule : public runtime::ModuleNode {
+ public:
+  /*!
+   * \brief Get member function to front-end
+   * \param name The name of the function.
+   * \param sptr_to_self The pointer to the module node.
+   * \return The corresponding member function.
+   */
+  virtual PackedFunc GetFunction(const std::string& name,
+                                 const std::shared_ptr<ModuleNode>& sptr_to_self) {
+    if (name == "get_graph_json") {
```

Comment:
maybe you should wrap this in a macro to simplify code. or,
you write a single return PackedFunc([name, ...](...) {
}, and inside you do this match. it wouldnt be a bottleneck.

Reply:
There are only handful functions, I don't think macro is helpful for maintaining.